documentation/proposals/Proposal - Silk.NET 3.X Generic Maths.md

```diff
@@ -0,0 +1,884 @@
+- Should try to roughly match System.Numerics APIs, with concessions for design oversights in those APIs
+- Should use vectorized implementations for methods when possible
+- Prefer generic specialization and extension methods, and keep the type of `T` as wide as possible in the struct itself
+- Primitives are read-only, `with` preferred for modification
+- Public fields instead of properties
+- Implement the following interfaces:
+    - IEquatable
+    - IEquatable to equivalent System.Numerics primitive, if available
+    - IReadOnlyList
+    - IUtf8SpanFormattable
+    - IUtf8SpanParsable
+- Provide explicit implementations of the following interfaces:
+    - INumberBase (except for methods which are relevant to vectors)
+- Provide implementations of the following interfaces, but as extension/static methods:
+    - IHyperbolicFunctions
+    - ITrigonometricFunctions
+    - ILogarithmicFunctions
+    - IExponentialFunctions
+    - IPowerFunctions
+    - IRootFunctions
+    (The reasoning for this is that the required interface types are more narrow than INumberBase)
+- Static helper methods, where the result is expected to be bounded to 0-1 or the type of T has to be scoped to a type narrower than the one used by the primitive, have two variants:
+    - Method&lt;T>: T is scoped to a narrower type (e.g IRootFunctions&lt;T>, IFloatingPoint&lt;T>)
+    - Method&lt;T, TReturn>: no scope for T, but TReturn is scoped to a narrower type
+
+# Proposed API
+- IVector&lt;TVector, T>
+- IVectorAlso&lt;TVector, T> - Covers interfaces that can't be implemented by IVector because of compiler restrictions
+- Vector2D&lt;T>
+- Vector3D&lt;T>
+- Vector4D&lt;T>
+- Quaternion&lt;T>
+- Matrix3x2&lt;T>
+- Matrix4x4&lt;T>
+- Plane&lt;T>
+
+For all `VectorND<T>`:
+```cs
+public readonly partial struct Vector2D<T> : IVector<Vector2D<T>, T>, IVectorAlso<Vector2D<T>, T>, IEquatable<Vector2>, IVec2
+    where T : INumberBase<T>
+{
+    /// <summary>The X component of the vector.</summary>
+    public readonly T X;
+
+    /// <summary>The Y component of the vector.</summary>
+    public readonly T Y;
+
+    internal const int Count = 2;
+
+    /// <summary>Creates a new <see cref="Vector2D{T}" /> object whose two elements have the same value.</summary>
+    /// <param name="value">The value to assign to all two elements.</param>
+    public Vector2D(T value);
+
+    /// <summary>Creates a vector whose elements have the specified values.</summary>
+    /// <param name="x">The value to assign to the <see cref="X" /> field.</param>
+    /// <param name="y">The value to assign to the <see cref="Y" /> field.</param>
+    public Vector2D(T x, T y);
+
+    /// <summary>Constructs a vector from the given <see cref="ReadOnlySpan{T}" />. The span must contain at least 4 elements.</summary>
+    /// <param name="values">The span of elements to assign to the vector.</param>
+    public Vector2D(ReadOnlySpan<T> values);
```

Comment:
nit: Constructors are nice and all, but the `System.Runtime.Intrinsics` APIs explicitly went with `Vector.Create` instead, both for performance reasons (avoids requiring `this` to be passed `byref` as is required for instance methods) and because it avoids unnecessary generic instantiations.

Reply:
Is the JIT not able to optimize the case of the constructor being inlined, so that `this` being by-reference doesn't have any effect? Also this matches 2.X API, and allows for `new()`, which is very convenient. Could be changed, though.

Reply:
It can, in many cases. But that’s additional work/complexity for the JIT to handle and it can’t always figure it out for more complex code.

Setting the code up in a way that is efficient and makes the JITs job easier is general goodness.

Reply:
new() works with any struct, because the default constructor always exists

Other constructors aren’t accessible from generic contexts and need named static methods, like Create or CreateSatursting/Truncating/Checked anyways

Reply:
> new() works with any struct, because the default constructor always exists

I specifically meant target-typed constructor invocations, which reduce a bit of friction when working with a contrived type like Vector3D&lt;float> or Matrix5X4&lt;Complex>. I'm aiming to ease the friction on the common case here, which is not TVector.

Reply:
type inference should “just work” with Vector2D.Create(…) too. it’s slightly longer than just new(…), but it works in more contexts and can be more efficient

you could always support both, and will need Create if you want generic contexts to allow creation at all. Really just makes a difference on if it’s implicit or explicit